File: drivers/java/src/main/java/dev/reddb/http/HttpConn.java
package dev.reddb.http;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import dev.reddb.Conn;
import dev.reddb.Options;
import dev.reddb.RedDBException;
import dev.reddb.Url;

import javax.net.ssl.SSLContext;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;

/**
 * HTTP transport. Mirrors the Rust / JS HTTP drivers: a single
 * {@link HttpClient} talking JSON to the RedDB REST endpoints,
 * carrying a bearer token in {@code Authorization}. Login is
 * automatic when {@link Options} carries username + password.
 */
public final class HttpConn implements Conn {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final HttpClient client;
    private final String baseUrl;
    private volatile String token;
    private final Duration timeout;
    private volatile boolean closed;

    public HttpConn(HttpClient client, String baseUrl, String token, Duration timeout) {
        this.client = client;
        this.baseUrl = stripTrailingSlash(baseUrl);
        this.token = token;
        this.timeout = timeout == null ? Duration.ofSeconds(30) : timeout;
    }

    /** Open a fresh client, log in if credentials were supplied, and return a ready connection. */
    public static HttpConn connect(Url url, Options opts) {
        String scheme = url.kind() == Url.Kind.HTTPS ? "https" : "http";
        String baseUrl = scheme + "://" + url.host() + ":" + url.port();
        HttpClient.Builder b = HttpClient.newBuilder()
            .version(HttpClient.Version.HTTP_1_1)
            .connectTimeout(opts.timeout());
        if (url.isTls()) {
            try {
                b.sslContext(SSLContext.getDefault());
            } catch (Exception e) {
                throw new RedDBException.ProtocolError("default SSLContext unavailable: " + e.getMessage(), e);
            }
        }
        HttpClient client = b.build();

        String token = opts.token() != null ? opts.token() : url.token();
        HttpConn conn = new HttpConn(client, baseUrl, token, opts.timeout());

        // Auto-login when the URL / options carry credentials but no token.
        if (token == null) {
            String user = opts.username() != null ? opts.username() : url.username();
            String pass = opts.password() != null ? opts.password() : url.password();
            if (user != null && pass != null) {
                conn.login(user, pass);
            }
        }
        return conn;
    }

    /** POST /auth/login → updates this connection's bearer token. */
    public void login(String username, String password) {
        ObjectNode body = MAPPER.createObjectNode();
        body.put("username", username);
        body.put("password", password);
        byte[] resp = post("/auth/login", body, false);
        try {
            JsonNode j = MAPPER.readTree(resp);
            JsonNode tok = j.get("token");
            if (tok == null || !tok.isTextual()) {
                // Some envelopes wrap as { ok, result: { token } }
                JsonNode inner = j.get("result");
                if (inner != null && inner.isObject()) tok = inner.get("token");
            }
            if (tok == null || !tok.isTextual()) {
                throw new RedDBException.ProtocolError("auth/login response missing 'token'");
            }
            this.token = tok.asText();
        } catch (IOException e) {
            throw new RedDBException.ProtocolError("auth/login: invalid JSON: " + e.getMessage(), e);
        }
    }

    @Override
    public byte[] query(String sql) {
        ObjectNode body = MAPPER.createObjectNode();
        body.put("sql", sql);
        return post("/query", body, true);
    }

    @Override
    public void insert(String collection, Object payload) {
        ObjectNode body = MAPPER.createObjectNode();
        body.put("collection", collection);
        body.set("payload", MAPPER.valueToTree(payload));
        post("/insert", body, true);
    }

    @Override
    public void bulkInsert(String collection, List<?> rows) {
        ObjectNode body = MAPPER.createObjectNode();
        body.put("collection", collection);
        body.set("payloads", MAPPER.valueToTree(rows));
        post("/bulk_insert", body, true);
    }

    @Override
    public byte[] get(String collection, String id) {
        ObjectNode body = MAPPER.createObjectNode();
        body.put("collection", collection);
        body.put("id", id);
        return post("/get", body, true);
    }

    @Override
    public void delete(String collection, String id) {
        ObjectNode body = MAPPER.createObjectNode();
        body.put("collection", collection);
        body.put("id", id);
        post("/delete", body, true);
    }

    @Override
    public void ping() {
        // GET /admin/health — anything 2xx counts as healthy.
        HttpRequest.Builder b = HttpRequest.newBuilder(URI.create(baseUrl + "/admin/health"))
            .timeout(timeout)
            .header("accept", "application/json")
            .GET();
        if (token != null) {
            b = b.header("authorization", "Bearer " + token);
        }
        try {
            HttpResponse<byte[]> resp = client.send(b.build(), HttpResponse.BodyHandlers.ofByteArray());
            if (resp.statusCode() / 100 != 2) {
                throw new RedDBException.EngineError(
                    "ping: HTTP " + resp.statusCode() + ": " + new String(resp.body(), StandardCharsets.UTF_8));
            }
        } catch (IOException | InterruptedException e) {
            if (e instanceof InterruptedException) Thread.currentThread().interrupt();
            throw new RedDBException.ProtocolError("ping I/O: " + e.getMessage(), e);
        }
    }

    @Override
    public void close() {
        // HttpClient is stateless; nothing to release.
        closed = true;
    }

    public String token() { return token; }
    public boolean isClosed() { return closed; }

    /** POST a JSON body and return the raw response bytes. */
    private byte[] post(String path, ObjectNode body, boolean requireAuthHeader) {
        try {
            byte[] payload = MAPPER.writeValueAsBytes(body);
            HttpRequest.Builder b = HttpRequest.newBuilder(URI.create(baseUrl + path))
                .timeout(timeout)
                .header("content-type", "application/json")
                .header("accept", "application/json")
                .POST(HttpRequest.BodyPublishers.ofByteArray(payload));
            if (requireAuthHeader && token != null) {
                b = b.header("authorization", "Bearer " + token);
            }
            HttpResponse<byte[]> resp = client.send(b.build(), HttpResponse.BodyHandlers.ofByteArray());
            int sc = resp.statusCode();
            byte[] respBody = resp.body();
            if (sc / 100 != 2) {
                String msg = respBody == null ? "" : new String(respBody, StandardCharsets.UTF_8);
                if (sc == 401 || sc == 403) {
                    throw new RedDBException.AuthRefused("HTTP " + sc + " " + path + ": " + msg);
                }
                throw new RedDBException.EngineError("HTTP " + sc + " " + path + ": " + msg);
            }
            return respBody;
        } catch (IOException e) {
            throw new RedDBException.ProtocolError(path + " I/O: " + e.getMessage(), e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RedDBException.ProtocolError(path + " interrupted", e);
        }
    }

    private static String stripTrailingSlash(String s) {
        if (s == null) return null;
        return s.endsWith("/") ? s.substring(0, s.length() - 1) : s;
    }
}

File: drivers/kotlin/src/main/kotlin/dev/reddb/http/HttpConn.kt
package dev.reddb.http

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import dev.reddb.Conn
import dev.reddb.Options
import dev.reddb.RedDBException
import dev.reddb.Url
import io.ktor.client.HttpClient
import io.ktor.client.engine.cio.CIO
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.readBytes
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.contentType
import java.nio.charset.StandardCharsets
import kotlin.time.Duration

/**
 * HTTP transport. Mirrors the JS / Java HTTP drivers: a single ktor
 * [HttpClient] (CIO engine) talking JSON to the RedDB REST endpoints
 * with a bearer token in `Authorization`. Login is automatic when
 * the URL or [Options] carries username + password.
 */
public class HttpConn internal constructor(
    private val client: HttpClient,
    baseUrl: String,
    @Volatile private var token: String?,
    private val timeout: Duration,
) : Conn {
    private val baseUrl: String = stripTrailingSlash(baseUrl)
    @Volatile private var closed: Boolean = false

    public companion object {
        internal val MAPPER: ObjectMapper = jacksonObjectMapper()

        /** Open a fresh client, auto-log-in when credentials are present. */
        public suspend fun connect(url: Url, opts: Options): HttpConn {
            val scheme = if (url.kind == Url.Kind.HTTPS) "https" else "http"
            val host = url.host ?: throw IllegalArgumentException("URL is missing a host")
            val baseUrl = "$scheme://$host:${url.port}"
            val timeoutMs = opts.timeout.inWholeMilliseconds

            val client = HttpClient(CIO) {
                expectSuccess = false
                engine {
                    requestTimeout = timeoutMs
                    endpoint.connectTimeout = timeoutMs
                    endpoint.socketTimeout = timeoutMs
                }
            }

            val token = opts.token ?: url.token
            val conn = HttpConn(client, baseUrl, token, opts.timeout)

            if (token == null) {
                val user = opts.username ?: url.username
                val pass = opts.password ?: url.password
                if (user != null && pass != null) {
                    conn.login(user, pass)
                }
            }
            return conn
        }
    }

    /** POST /auth/login → updates this connection's bearer token. */
    public suspend fun login(username: String, password: String) {
        val body = MAPPER.createObjectNode().apply {
            put("username", username)
            put("password", password)
        }
        val resp = post("/auth/login", body, requireAuthHeader = false)
        try {
            val j = MAPPER.readTree(resp)
            var tok = j.get("token")
            if (tok == null || !tok.isTextual) {
                val inner = j.get("result")
                if (inner != null && inner.isObject) tok = inner.get("token")
            }
            if (tok == null || !tok.isTextual) {
                throw RedDBException.ProtocolError("auth/login response missing 'token'")
            }
            this.token = tok.asText()
        } catch (e: RedDBException) {
            throw e
        } catch (e: Exception) {
            throw RedDBException.ProtocolError("auth/login: invalid JSON: ${e.message}", e)
        }
    }

    override suspend fun query(sql: String): ByteArray {
        val body = MAPPER.createObjectNode().put("sql", sql)
        return post("/query", body, requireAuthHeader = true)
    }

    override suspend fun insert(collection: String, payload: Any) {
        val body = MAPPER.createObjectNode().apply {
            put("collection", collection)
            set<com.fasterxml.jackson.databind.JsonNode>("payload", MAPPER.valueToTree(payload))
        }
        post("/insert", body, requireAuthHeader = true)
    }

    override suspend fun bulkInsert(collection: String, rows: List<Any?>) {
        val body = MAPPER.createObjectNode().apply {
            put("collection", collection)
            set<com.fasterxml.jackson.databind.JsonNode>("payloads", MAPPER.valueToTree(rows))
        }
        post("/bulk_insert", body, requireAuthHeader = true)
    }

    override suspend fun get(collection: String, id: String): ByteArray {
        val body = MAPPER.createObjectNode().apply {
            put("collection", collection)
            put("id", id)
        }
        return post("/get", body, requireAuthHeader = true)
    }

    override suspend fun delete(collection: String, id: String) {
        val body = MAPPER.createObjectNode().apply {
            put("collection", collection)
            put("id", id)
        }
        post("/delete", body, requireAuthHeader = true)
    }

    override suspend fun ping() {
        try {
            val resp: HttpResponse = client.get("$baseUrl/admin/health") {
                header(HttpHeaders.Accept, "application/json")
                if (token != null) header(HttpHeaders.Authorization, "Bearer $token")
            }
            val sc = resp.status.value
            if (sc / 100 != 2) {
                val body = String(resp.readBytes(), StandardCharsets.UTF_8)
                throw RedDBException.EngineError("ping: HTTP $sc: $body")
            }
        } catch (e: RedDBException) {
            throw e
        } catch (e: Throwable) {
            throw RedDBException.ProtocolError("ping I/O: ${e.message}", e)
        }
    }

    override fun close() {
        if (closed) return
        closed = true
        try { client.close() } catch (_: Throwable) { /* ignore */ }
    }

    public fun token(): String? = token
    public fun isClosed(): Boolean = closed

    /** POST a JSON body and return the raw response bytes. */
    private suspend fun post(path: String, body: ObjectNode, requireAuthHeader: Boolean): ByteArray {
        try {
            val payload = MAPPER.writeValueAsBytes(body)
            val resp: HttpResponse = client.post("$baseUrl$path") {
                contentType(ContentType.Application.Json)
                header(HttpHeaders.Accept, "application/json")
                if (requireAuthHeader && token != null) {
                    header(HttpHeaders.Authorization, "Bearer $token")
                }
                setBody(payload)
            }
            val sc = resp.status.value
            val respBody = resp.readBytes()
            if (sc / 100 != 2) {
                val msg = String(respBody, StandardCharsets.UTF_8)
                if (sc == 401 || sc == 403) {
                    throw RedDBException.AuthRefused("HTTP $sc $path: $msg")
                }
                throw RedDBException.EngineError("HTTP $sc $path: $msg")
            }
            return respBody
        } catch (e: RedDBException) {
            throw e
        } catch (e: Throwable) {
            throw RedDBException.ProtocolError("$path I/O: ${e.message}", e)
        }
    }

    private fun stripTrailingSlash(s: String): String =
        if (s.endsWith("/")) s.substring(0, s.length - 1) else s
}
